File: frontend/js/sse.js
/**
 * Server-Sent Events (SSE) Connection Handler
 * 
 * Manages real-time one-way connection from backend to frontend.
 * Receives sensor data updates via SSE and notifies callback functions.
 * 
 * Benefits over WebSocket:
 * - Simpler protocol (one-way only)
 * - Automatic reconnection built into browser
 * - Less overhead
 * - Better for server → client streaming
 * 
 * Backend sends data as JSON array of sensor readings:
 * [{temp: 22.5, hum: 45.0, motion: false, sound_db: 35.2, deviceid: "pi-001", timestamp: "..."}]
 */

class SSEManager {
    constructor() {
        this.eventSource = null;
        this.callbacks = [];
        this.statusCallback = null;
        this.reconnectAttempts = 0;
        this.maxReconnectAttempts = 10;
    }

    /**
     * Connect to SSE stream
     * @param {string} url - SSE URL (uses CONFIG.SSE_URL by default)
     */
    connect(url = CONFIG.SSE_URL) {
        try {
            // Close existing connection if any
            if (this.eventSource) {
                this.eventSource.close();
            }

            this.eventSource = new EventSource(url);
            
            this.eventSource.onopen = () => {
                console.log('[SSE] Connected to backend');
                this.updateStatus(true);
                this.reconnectAttempts = 0; // Reset counter on successful connection
            };

            this.eventSource.onmessage = (event) => {
                try {
                    // Parse incoming JSON data
                    const data = JSON.parse(event.data);
                    
                    // Notify all registered callbacks
                    this.callbacks.forEach(callback => {
                        callback(data);
                    });
                } catch (error) {
                    console.error('[SSE] Error parsing data:', error);
                }
            };

            this.eventSource.onerror = (error) => {
                console.error('[SSE] Connection error');
                this.updateStatus(false);
                
                // EventSource automatically attempts to reconnect
                // but we track attempts to show user if persistent failure
                this.reconnectAttempts++;
                
                if (this.reconnectAttempts >= this.maxReconnectAttempts) {
                    console.error('[SSE] Max reconnection attempts reached');
                    this.eventSource.close();
                }
            };

        } catch (error) {
            console.error('[SSE] Failed to connect:', error);
            this.updateStatus(false);
        }
    }

    /**
     * Register callback for data updates
     * @param {function} callback - Function to call with new data
     */
    onData(callback) {
        this.callbacks.push(callback);
    }

    /**
     * Register callback for status updates
     * @param {function} callback - Function to call with connection status (true/false)
     */
    onStatus(callback) {
        this.statusCallback = callback;
    }

    /**
     * Update connection status and notify callback
     */
    updateStatus(connected) {
        if (this.statusCallback) {
            this.statusCallback(connected);
        }
    }

    /**
     * Close SSE connection
     */
    close() {
        if (this.eventSource) {
            this.eventSource.close();
            this.eventSource = null;
        }
        this.reconnectAttempts = 0;
    }
}

// Export singleton instance
const sseManager = new SSEManager();

File: backend/src/sse.rs
use axum::{
    extract::State,
    response::{sse::Event, Sse},
};
use futures::stream::{self, Stream};
use redis::aio::ConnectionManager;
use std::convert::Infallible;
use std::time::Duration;

use crate::models::sensor_data::SensorData;
use crate::routes::sensor_data::AppState;

/// Real-time state for SSE streaming (Branch 1)
/// 
/// This structure manages Redis cache for the last 100 sensor readings,
/// enabling zero-latency real-time monitoring for the frontend via SSE.
/// 
/// Architecture (Branch 1 - Real-time Path):
/// Pi → Backend → Redis Cache → SSE → Frontend
/// 
/// Why SSE instead of WebSocket?
/// - One-way communication (sensors → frontend only)
/// - Simpler protocol, less overhead
/// - Automatic reconnection built into browser
/// - Better for streaming updates from server
/// 
/// Redis Data Structure:
/// - Key: "sensor:latest"
/// - Type: List (LPUSH for new data, LTRIM to keep last 100)
/// - TTL: 2 hours (auto-expiry if no updates)
#[derive(Clone)]
pub struct RealtimeState {
    /// Redis connection manager for async operations
    pub redis: ConnectionManager,
}

impl RealtimeState {
    /// Create new RealtimeState with Redis connection
    pub async fn new(redis_url: &str) -> Result<Self, redis::RedisError> {
        let client = redis::Client::open(redis_url)?;
        let redis = ConnectionManager::new(client).await?;
        Ok(Self { redis })
    }
}

/// SSE handler - streams sensor data to client
/// 
/// GET /events
/// 
/// This endpoint establishes a Server-Sent Events connection,
/// allowing one-way real-time communication from server to frontend.
/// 
/// Flow:
/// 1. Client sends HTTP GET request to /events
/// 2. Server responds with Content-Type: text/event-stream
/// 3. Server continuously sends data events
/// 4. Connection stays open until client disconnects
/// 
/// Example JavaScript client:
/// ```javascript
/// const eventSource = new EventSource('http://localhost:3000/events');
/// eventSource.onmessage = (event) => {
///   const data = JSON.parse(event.data);
///   console.log('Received:', data);
/// };
/// ```
#[tracing::instrument(skip_all)]
pub async fn sse_handler(
    State(state): State<AppState>,
) -> Sse<impl Stream<Item = Result<Event, Infallible>>> {
    tracing::info!("SSE connection established");
    
    let stream = stream::unfold(
        (state, 0_u64, tokio::time::Instant::now()),
        move |(state, mut message_count, mut last_full_sync)| async move {
            // Hybrid approach: instant notifications + periodic full sync
            let mut rx = state.broadcast_tx.subscribe();
            
            tokio::select! {
                // Branch 1: New data notification (instant, zero latency)
                Ok(new_data) = rx.recv() => {
                    tracing::debug!("SSE: Sending instant notification");
                    
                    match serde_json::to_string(&vec![new_data]) {
                        Ok(json) => {
                            message_count += 1;
                            
                            if message_count % 60 == 0 {
                                tracing::debug!("SSE: sent {} instant updates", message_count);
                            }
                            
                            Some((
                                Ok(Event::default().data(json)),
                                (state, message_count, last_full_sync)
                            ))
                        }
                        Err(e) => {
                            tracing::error!("Failed to serialize SSE data: {}", e);
                            None
                        }
                    }
                }
                
                // Branch 2: Periodic full sync (every 5 seconds as fallback)
                _ = tokio::time::sleep_until(last_full_sync + Duration::from_secs(5)) => {
                    tracing::debug!("SSE: Performing periodic full sync");
                    
                    // Read last 100 readings from Redis
                    let readings_result: Result<Vec<String>, redis::RedisError> = redis::cmd("LRANGE")
                        .arg("sensor:latest")
                        .arg(0)
                        .arg(99)
                        .query_async(&mut state.realtime.redis.clone())
                        .await;
                    
                    match readings_result {
                        Ok(readings_json) => {
                            let mut data: Vec<SensorData> = Vec::new();
                            for json_str in readings_json {
                                if let Ok(sensor_data) = serde_json::from_str::<SensorData>(&json_str) {
                                    data.push(sensor_data);
                                }
                            }
                            
                            match serde_json::to_string(&data) {
                                Ok(json) => {
                                    last_full_sync = tokio::time::Instant::now();
                                    tracing::debug!("SSE: Full sync sent - {} readings", data.len());
                                    
                                    Some((
                                        Ok(Event::default().data(json)),
                                        (state, message_count, last_full_sync)
                                    ))
                                }
                                Err(e) => {
                                    tracing::error!("Failed to serialize full sync: {}", e);
                                    None
                                }
                            }
                        }
                        Err(e) => {
                            tracing::error!("Redis error during full sync: {}", e);
                            None
                        }
                    }
                }
            }
        },
    );
    
    Sse::new(stream).keep_alive(
        axum::response::sse::KeepAlive::new()
            .interval(Duration::from_secs(15))
            .text("keep-alive")
    )
}
